File: swifty_crypto/envelope.py
"""Encryption envelope (contracts/identity-envelope.md §4): XChaCha20-Poly1305 AEAD.

The server stores only opaque `envelope` dicts; it holds no key and cannot decrypt. The
AAD binds the blob to account+purpose+version, so a blob can't be replayed across users or
purposes (the AEAD tag covers the AAD; tampering it fails the open).
"""

from __future__ import annotations

import hashlib
import hmac

import nacl.bindings as sodium
import nacl.utils

ALG = "xchacha20poly1305"
NONCE_LEN = sodium.crypto_aead_xchacha20poly1305_ietf_NPUBBYTES  # 24


def _aad(account_id: str, purpose: str, v: int) -> bytes:
    # length-delimited (newline) so the fields can't be re-segmented ambiguously (B1).
    return hashlib.sha256(f"{account_id}\n{purpose}\n{v}".encode()).digest()


def seal(key: bytes, plaintext: bytes, *, account_id: str, purpose: str, v: int, kid: str = "v1") -> dict:
    nonce = nacl.utils.random(NONCE_LEN)  # 24 random bytes per encryption (B4)
    aad = _aad(account_id, purpose, v)
    ct = sodium.crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, aad, nonce, key)
    return {"v": v, "alg": ALG, "kid": kid, "nonce": nonce, "aad": aad, "ct": ct}


def open(  # noqa: A001 - contract API name (identity-envelope §4)
    key: bytes,
    envelope: dict,
    *,
    account_id: str | None = None,
    purpose: str | None = None,
) -> bytes:
    """Decrypt + verify. Raises on any tamper — never returns data for a modified
    ciphertext or AAD.

    Callers SHOULD pass `account_id` + `purpose`: when given, the stored AAD is
    re-derived from those authoritative inputs and checked, so the §4 user+purpose+version
    binding is enforced on READ (not just trusted from the stored field). Omitting them
    (the bare contract `open(key, envelope)` form) decrypts against the stored AAD only."""
    aad = envelope["aad"]
    if account_id is not None and purpose is not None:
        expected = _aad(account_id, purpose, envelope["v"])
        if not hmac.compare_digest(expected, aad):
            raise ValueError("AAD does not bind to the expected account_id/purpose/v")
        aad = expected
    return sodium.crypto_aead_xchacha20poly1305_ietf_decrypt(
        envelope["ct"], aad, envelope["nonce"], key
    )

File: caladon-core/src/envelope.rs
//! Encryption envelope (contracts/identity-envelope.md §4): XChaCha20-Poly1305-IETF AEAD.
//! Byte-identical to `swifty_crypto/envelope.py` (the reference) and `SwiftyKit/Crypto.swift`.
//!
//! The server stores only opaque envelopes; it holds no key and cannot decrypt. The AAD binds
//! the blob to account+purpose+version (B1), so a blob can't be replayed across users or
//! purposes — the AEAD tag covers the AAD, so tampering it fails the open (fail-closed).
//!
//! `chacha20poly1305::XChaCha20Poly1305` is the libsodium `crypto_aead_xchacha20poly1305_ietf`
//! construction: 24-byte (192-bit) nonce, 16-byte Poly1305 tag appended to the ciphertext —
//! exactly what pynacl's `crypto_aead_xchacha20poly1305_ietf_encrypt` returns.

use chacha20poly1305::aead::{Aead, AeadCore, KeyInit, OsRng, Payload};
use chacha20poly1305::{Key, XChaCha20Poly1305, XNonce};
use sha2::{Digest, Sha256};

pub const ALG: &str = "xchacha20poly1305";
/// XChaCha20-Poly1305-IETF NPUBBYTES (== libsodium crypto_aead_xchacha20poly1305_ietf_NPUBBYTES).
pub const NONCE_LEN: usize = 24;
pub const KEY_LEN: usize = 32;
/// Poly1305 tag length (libsodium ABYTES); appended to the ciphertext by the IETF construction.
pub const TAG_LEN: usize = 16;

#[derive(Debug, PartialEq, Eq)]
pub enum EnvelopeError {
    BadKeyLength,
    BadNonceLength,
    /// Decrypt/verify failed — ciphertext or AAD tampered, or wrong key/nonce.
    OpenFailed,
    SealFailed,
}

/// Envelope AAD = SHA-256("{account_id}\n{purpose}\n{v}") — byte-identical to
/// `swifty_crypto._aad` (newline-delimited so the fields can't be re-segmented; B1).
pub fn aad(account_id: &str, purpose: &str, v: i64) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(format!("{account_id}\n{purpose}\n{v}").as_bytes());
    h.finalize().into()
}

/// Deterministic seal with a caller-supplied 24-byte nonce — the parity entry point (Python's
/// `seal` draws a random nonce, so this mirrors the inner AEAD call for fixed-vector testing).
/// Returns the ciphertext+tag (`ct`), exactly as pynacl's
/// `crypto_aead_xchacha20poly1305_ietf_encrypt(plaintext, aad, nonce, key)`.
pub fn seal_with_nonce(
    key: &[u8],
    plaintext: &[u8],
    account_id: &str,
    purpose: &str,
    v: i64,
    nonce: &[u8],
) -> Result<Vec<u8>, EnvelopeError> {
    if key.len() != KEY_LEN {
        return Err(EnvelopeError::BadKeyLength);
    }
    if nonce.len() != NONCE_LEN {
        return Err(EnvelopeError::BadNonceLength);
    }
    let ad = aad(account_id, purpose, v);
    seal_raw(key, plaintext, &ad, nonce)
}

/// Low-level seal: AAD passed in directly (used by `session`, which derives its own AAD).
pub fn seal_raw(
    key: &[u8],
    plaintext: &[u8],
    ad: &[u8],
    nonce: &[u8],
) -> Result<Vec<u8>, EnvelopeError> {
    if key.len() != KEY_LEN {
        return Err(EnvelopeError::BadKeyLength);
    }
    if nonce.len() != NONCE_LEN {
        return Err(EnvelopeError::BadNonceLength);
    }
    let cipher = XChaCha20Poly1305::new(Key::from_slice(key));
    cipher
        .encrypt(
            XNonce::from_slice(nonce),
            Payload {
                msg: plaintext,
                aad: ad,
            },
        )
        .map_err(|_| EnvelopeError::SealFailed)
}

/// Random-nonce seal (production path). Returns `(nonce, ct)`. A fresh CSPRNG 24-byte nonce per
/// call (identity-envelope §4/B4). Mirrors `swifty_crypto.seal` minus the dict wrapper.
pub fn seal(
    key: &[u8],
    plaintext: &[u8],
    account_id: &str,
    purpose: &str,
    v: i64,
) -> Result<([u8; NONCE_LEN], Vec<u8>), EnvelopeError> {
    let nonce = XChaCha20Poly1305::generate_nonce(&mut OsRng);
    let ct = seal_with_nonce(key, plaintext, account_id, purpose, v, nonce.as_slice())?;
    let mut n = [0u8; NONCE_LEN];
    n.copy_from_slice(nonce.as_slice());
    Ok((n, ct))
}

/// Decrypt + verify against the supplied AAD. Fails closed on ANY tamper (ct or AAD) — never
/// returns data for a modified envelope. Mirrors `swifty_crypto.open` / `SwiftyCrypto.open`.
pub fn open(key: &[u8], nonce: &[u8], ad: &[u8], ct: &[u8]) -> Result<Vec<u8>, EnvelopeError> {
    if key.len() != KEY_LEN {
        return Err(EnvelopeError::BadKeyLength);
    }
    if nonce.len() != NONCE_LEN {
        return Err(EnvelopeError::BadNonceLength);
    }
    let cipher = XChaCha20Poly1305::new(Key::from_slice(key));
    cipher
        .decrypt(
            XNonce::from_slice(nonce),
            Payload {
                msg: ct,
                aad: ad,
            },
        )
        .map_err(|_| EnvelopeError::OpenFailed)
}

/// Convenience open that re-derives the AAD from account_id/purpose/v (enforces the §4 binding
/// on READ, matching Python's `open(..., account_id=, purpose=)` form).
pub fn open_bound(
    key: &[u8],
    nonce: &[u8],
    ct: &[u8],
    account_id: &str,
    purpose: &str,
    v: i64,
) -> Result<Vec<u8>, EnvelopeError> {
    let ad = aad(account_id, purpose, v);
    open(key, nonce, &ad, ct)
}
